### PR TITLE
12-Add-support-for-backups-and-connection-statement-cursor-division-of-responsibilities

### PR DESCRIPTION
**Pull request by pharo-db**
Added support for backups and clarified responsibilities or Connection, Statement, and Cursor.

### CODE REVIEW

**Comment by tblanchard** [comment]
Addresses Issue #12.

File: src/SQLite3-Core/SQLite3Backup.class.st [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+		'srcName',
+		'dstConnection',
+		'dstName',
+		'isOpen',
+		'handle'
+	],
+	#category : #'SQLite3-Core-Connections'
+}
+
+{ #category : #'as yet unclassified' }
```
Comment (astares) [inline src/SQLite3-Core/SQLite3Backup.class.st]: should be classified

File: src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st [hunk 2]
```diff
@@ -442,7 +473,7 @@ SQLiteConnectionTest >> testExecuteWithInsertBooleanSQLNullThenSelect [
 	self assert: res equals: 0.
 
 	rs := db execute: 'insert into x values (NULL, NULL)'.
-	self assert: rs next isNil.
+	self assert: rs next equals: nil.
```
Comment (astares) [inline src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st]: Why use the #assert:equals: here when using the #assert: with #isNil checking for a boolean is fine too?

File: src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st [hunk 3]
```diff
@@ -48,9 +49,39 @@ SQLiteConnectionTest >> setUp [
...
+{ #category : #tests }
+SQLiteConnectionTest >> testBackup [
+	| row backup |
+	
+	target := SQLite3Connection memory.
+	target open.
+	self populateRowsInto: 'junk' of: db.
+	backup := db backupTo: target.
+	backup open; step; finish.
+	row := (target execute: 'select * from junk where rt = ''1-ZZZ''') onlyRow.
```
Comment (astares) [inline src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st]: This is testing the target not the backup. No?

File: src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st (hunk 2 above)
Comment (pharo-db) [reply src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st]: No good reason other than I had changed it to return something else for a bit and had not put it back exactly the same.  I will revert that.

File: src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st (hunk 3 above)
Comment (pharo-db) [reply src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st]: The backup is a backup object - it is a process kind of object like a statement.  It has step's that you can repeatedly call to push the backup process forwards.  You could implement a progress indicator with it.

The target is the database we are copying db to.  If target has the same data as db, then the backup succeeded.

Comment (pharo-db) [reply src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st]: sqlite3_backup* sqlite3_backup_init(
                    sqlite3* db_dst, const char* db_name_dst,
                    sqlite3* db_src, const char* db_name_src );

backup is the pharo object representing sqlite3_backup.

Check out https://www.oreilly.com/library/view/using-sqlite/9781449394592/re225.html

File: src/SQLite3-Core/SQLite3Backup.class.st (hunk 1 above)
Comment (pharo-db) [reply src/SQLite3-Core/SQLite3Backup.class.st]: agree, will fix that.

**Comment by astares** [comment]
CI Tests fail due to missing method:

SQLite3RowTest #testAsDictionary is failing in 6.1: MessageNotUnderstood: Dictionary class>>newFromKeys:andValues:

**Comment by pharo-db** [comment]
I see, I suppose that is new in Pharo 8?  I will do a workaround.

File: src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st (hunk 2 above)
Comment (pharo-db) [reply src/SQLite3-Core-Tests/SQLiteConnectionTest.class.st]: Reverted

**Pull request merged**